android/java/org/chromium/chrome/browser/app/BraveActivity.java

```diff
@@ -203,13 +236,75 @@ public boolean onMenuOrKeyboardAction(int id, boolean fromMenu) {
             openNewOrSelectExistingTab(REWARDS_SETTINGS_URL);
         } else if (id == R.id.brave_wallet_id) {
             openBraveWallet();
+        } else if (id == R.id.request_brave_vpn_id || id == R.id.request_brave_vpn_check_id) {
+            BraveVpnUtils.showProgressDialog(BraveActivity.this);
+            if (BraveVpnProfileUtils.getInstance(BraveActivity.this).isVPNConnected()) {
+                BraveVpnProfileUtils.getInstance(BraveActivity.this).stopVpn();
+            } else {
+                if (BraveVpnPrefUtils.isSubscriptionPurchase()) {
+                    verifySubscription();
+                } else {
+                    BraveVpnUtils.dismissProgressDialog();
+                    BraveVpnUtils.openBraveVpnPlansActivity(BraveActivity.this);
+                }
+            }
         } else {
             return false;
         }
 
         return true;
     }
 
+    private void verifySubscription() {
+        List<Purchase> purchases = InAppPurchaseWrapper.getInstance().queryPurchases();
+        if (purchases.size() == 1) {
+            Purchase purchase = purchases.get(0);
+            mPurchaseToken = purchase.getPurchaseToken();
+            mProductId = purchase.getSkus().get(0).toString();
+            BraveVpnNativeWorker.getInstance().verifyPurchaseToken(mPurchaseToken, mProductId,
+                    BraveVpnUtils.SUBSCRIPTION_PARAM_TEXT, getPackageName());
+        } else {
+            braveVpnVerificationFailed();
+        }
+    }
+
+    @Override
+    public void onVerifyPurchaseToken(String jsonResponse, boolean isSuccess) {
+        if (isSuccess) {
+            Long purchaseExpiry = BraveVpnUtils.getPurchaseExpiryDate(jsonResponse);
+            if (purchaseExpiry > 0 && purchaseExpiry >= System.currentTimeMillis()) {
```

Comment:
With this wouldn't the users be able to just set their system clock back to a date earlier than the expiry time and bypass this check?

Reply:
@jumde you are right but it's implemented for the edge case only. we are requesting for purchase information from play store api and if the purchase is active then only it would return the response. it may be an issue for 5 minutes but we can't do anything here.